readthedocs/doc_builder/director.py

```diff
@@ -347,21 +285,26 @@ def build_epub(self):
         if "epub" not in self.data.config.formats or self.data.version.type == EXTERNAL:
             return False
 
-        commands = []  # self.data.config.build.jobs.build.epub
-        if commands:
-            for command in commands:
-                self.build_environment.run(command)
-            return True
-
         # Mkdocs has no epub generation currently.
         if self.is_type_sphinx():
             return self.build_docs_class("sphinx_epub")
         return False
 
-    def post_build(self):
-        commands = []  # self.data.config.build.jobs.post_build
+    def run_build_job(self, job):
+        if (
+            getattr(self.data.config.build, "jobs", None) is None
+            or getattr(self.data.config.build.jobs, job, None) is None
+        ):
+            return
+
+        cwd = self.data.project.checkout_path(self.data.version.slug)
+        environment = self.vcs_environment
+        if job not in ("pre_checkout", "post_checkout"):
+            environment = self.build_environment
+
+        commands = getattr(self.data.config.build.jobs, job, [])
         for command in commands:
-            self.build_environment.run(command)
+            environment.run(*command.split(), escape_command=False, cwd=cwd)
```

Comment:
This scares me a little bit, but we're already running Python inside these environments, so it's not meaningfully different.

Reply:
Wonder if we could save the state on each run, users will assume the `cwd` or environment variables to be on the same state as the previous command, like:

```
- cd foo
- FOO=bar
- pwd
- echo $FOO
```

Reply:
Hrm, I haven't thought too much about this. I don't think this is like an interactive bash terminal that keeps the state, tho. I'd be curious about how other CI work. Do you know if you can do this in GitHub Actions or CircleCI?

I think it won't be trivial to implement this, tho. In particular, considering how we are executing the commands in Docker since it spins a new bash per command.

Reply:
> Do you know if you can do this in GitHub Actions or CircleCI?

In circle we have https://github.com/readthedocs/readthedocs.org/blob/a7e30831f282b704a127290a0a235267eb638b39/.circleci/config.yml#L46-L50

but we should be able to do a quick test with the commands I pun in the example.

Reply:
I don't believe a list of commands will have any chance to persist state, not at least without writing to disk or something.

Docker's exec executes each command as an individual process:

![image](https://user-images.githubusercontent.com/1140183/158938061-6ea13abd-327a-4bd2-9c5f-cf04537ca482.png)

And our code also executes each `run()` in an individual shell process too:
https://github.com/readthedocs/readthedocs.org/blob/main/readthedocs/doc_builder/environments.py#L363-L368

So I would not expect environment variables and friends to persist between commands.

Now, chaining commands together would however execute them all in the same `run()` and same process:

```
- cd foo; FOO=bar; pwd; echo $FOO
```

This is executed as essentially:

`/bin/sh -c "cd foo; FOO=bar; pwd; echo $FOO"`

And this is also why the CircleCI execution works. Those aren't executed as individual commands, but as a singular shell script

Reply:
> This scares me a little bit

I'm sufficiently scared by arbitrary execution in general. I opened up a discussion on implementing gvisor a while back and I think it's still worth exploring.

The gvisor repo explains why containers are not sufficient security against arbitrary code:
https://github.com/google/gvisor

The gvisor integration is actually a pretty small change:
https://gvisor.dev/docs/user_guide/quick_start/docker/

Reply:
@agjohnson 

> Now, chaining commands together would however execute them all in the same run() and same process:

I'm fine coming back later and adding this as an improvement/feature in the next iteration if we find it's required/useful.

Reply:
> I'm fine coming back later

Same. I'm not advocating for any implementation here, just explaining the shell sequence we have in place now. I'd agree it's not a concern for this initial release.

I bet users can probably hack this together if it's necessary, or could write to disk if really necessary.